sensor/debugger/central/communicate_server.go

```diff
@@ -0,0 +1,122 @@
+package central
+
+import (
+	"testing"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
+	"google.golang.org/grpc/metadata"
+)
+
+// FakeService represents a fake central gRPC that reads and sends messages to sensor's connected gRPC stream.
+type FakeService struct {
+	ConnectionStarted concurrency.Signal
+	KillSwitch        concurrency.Signal
+
+	stream central.SensorService_CommunicateServer
+
+	// initialMessages are messages to be sent to sensor once connection is open
+	initialMessages []*central.MsgToSensor
+
+	// ordered messages received from sensor
+	receivedMessages []*central.MsgFromSensor
+
+	receivedLock sync.RWMutex
+
+	messageCallback func(sensor *central.MsgFromSensor)
+
+	t *testing.T
+}
+
+// GetAllMessages clones and returns all messages that were ingested by central gRPC.
+func (s *FakeService) GetAllMessages() []*central.MsgFromSensor {
+	var output []*central.MsgFromSensor
+	s.receivedLock.RLock()
+	defer s.receivedLock.RUnlock()
+	for _, m := range s.receivedMessages {
+		output = append(output, m.Clone())
+	}
+	return output
+}
+
+// ClearReceivedBuffer will wipe all messages read from sensor.
+func (s *FakeService) ClearReceivedBuffer() {
+	s.receivedLock.Lock()
+	defer s.receivedLock.Unlock()
+	s.receivedMessages = []*central.MsgFromSensor{}
+}
+
+// MakeFakeCentralWithInitialMessages creates a fake gRPC connection that sends `initialMessages` on startup.
+// Once communicate is called and the gRPC stream is enabled, this instance will send all `initialMessages` in order.
+func MakeFakeCentralWithInitialMessages(initialMessages ...*central.MsgToSensor) *FakeService {
+	return &FakeService{
+		ConnectionStarted: concurrency.NewSignal(),
+		KillSwitch:        concurrency.NewSignal(),
+		initialMessages:   initialMessages,
+		receivedMessages:  []*central.MsgFromSensor{},
+		receivedLock:      sync.RWMutex{},
+	}
+}
+
+// BlockReceive will read from the connection stream if the connection was already established.
+// The caller might get blocked because it waits until connection is available.
+func (s *FakeService) BlockReceive() (*central.MsgFromSensor, error) {
+	s.ConnectionStarted.Wait()
+	return s.stream.Recv()
+}
+
+// BlockSend will send from the connection stream if the connection was already established.
```

Comment:
Question: Is it *"will send __TO__ the connection stream"* or does `BlockSend` really send messages it takes from the connection stream and sends it somewhere?

Reply:
It was a typo, it sends a message to the stream. This is the fake central implementation file. Meaning that if you call `BlockSend`, sensor will receive a message.